### ISSUE TITLE
Using closures instead of function pointers?

### ISSUE THREAD

**Issue by Mek101** [glandium/os-thread-local#3]
Would it be possible to use closures instead of function pointers for `new()`?

**Comment by glandium** [comment]
Closures that don't borrow should coerce to functions, and those that do borrow would need the borrowed stuff to be Sync+Send. I don't think that makes for a great usecase.

**Comment by Mek101** [comment]
>  and those that do borrow would need the borrowed stuff to be Sync+Send

Do they though? If the data is initialized by the calling thread, there shouldn't be any concurrency issue, in particular, if the initialization is done in place with a closure provided by the calling thread

**Comment by glandium** [comment]
Let's take a step back. What is the kind of code you would like to put in a closure that does not coerce to a function?

**Comment by Mek101** [comment]
I wanted to share a `Hasher` from `https://github.com/abonander/img_hash/` between different `rayon` working threads, but apparently the struct isn't `Send` because there is some static global structure every instance shares: my solution was to make a wrapper where each thread has it's own thread local instance.
The fact is that I still wanted to be able to configure the `Hasher` by passing arguments to the `HasherConfig` from which the `Hasher` is build, and to do that I need a closure.

**Comment by glandium** [comment]
I don't see why you need a closure. You can have a static `HasherConfig`, or create the `HasherConfig` in the same function that creates the `Hasher`.

**Comment by Mek101** [comment]
> You can have a static HasherConfig

Not really, because I wanted to be able to have multiple instances differently configured

> create the HasherConfig in the same function that creates the Hasher.

But then I need to pass at an `alg` enum to the `HasherConfig`, which brings me the same issue
